### PR TITLE
Add field specific and general search to find command

### LINKED ISSUE #7: Finding family contacts using tags
I want to view all my contacts that are tagged “family”.

As a user, I should be able to find contacts that are tagged “family”.

### PR DESCRIPTION
**Pull request by darkdestry-t**
Checklist before exitting draft:

- [x] Code changes
- [x] Documentation
- [x] Test

Closes #7 #3 #91

@tanboonji Docs
@oeiyiping Model/Commands
@yaowei-soc Test

### CODE REVIEW

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 1]
```diff
@@ -65,4 +67,22 @@ public static boolean isNonZeroUnsignedInteger(String s) {
             return false;
         }
     }
+
+    /**
+     * Returns the similarity ratio of case insensitive fuzzy search using partial ratio if query is shorter and
+     * using simple ration if data is shorter.
```
Comment (tanboonji) [inline src/main/java/seedu/address/commons/util/StringUtil.java:73]: ```ratio``` spelled as ```ration```.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -25,9 +34,48 @@ public FindCommand parse(String args) throws ParseException {
...
+            if (remarkValue.isPresent()) {
+                RemarkContainsKeywordsPredicate predicateComparator =
+                        new RemarkContainsKeywordsPredicate(Arrays.asList(remarkValue.get().split("\\s+").clone()));
+                return new FindCommand(predicateComparator, predicateComparator);
+            }
+
+
+        }
+
+
```
Comment (tanboonji) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:75]: Extra line breaks in lines 71-72 and 74-75.

**Review by tanboonji — CHANGES_REQUESTED** [review]
Two minor changes requested. Other than that, ```Logic``` and ```Docs``` LGTM.

**Review by tanboonji — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        return o2Similarity - o1Similarity == 0
+                ? o1.getEmail().value.compareTo(o2.getEmail().value)
+                : o2Similarity - o1Similarity;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof EmailContainsKeywordsPredicate // instanceof handles nulls
+                && keywords.equals(((EmailContainsKeywordsPredicate) other).keywords)); // state check
```
Comment (yaowei-soc) [inline src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java:60]: Hmm using `.equals` would mean that ["first", "second"] would not be equals to ["second", "first"]. Is this an intended feature?

**Review by yaowei-soc — COMMENTED** [review]
Changes look good to me, however I have a pending question on the *KeywordsPredicate methods:
- EmailContainsKeywordsPredicate
- EmailContainsKeywordsPredicate
- RemarkContainsKeywordsPredicate
- TagContainsKeywordsPredicate

**Review by oeiyiping — APPROVED** [review]
Aside from the existing comments, LGTM.

File: src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java (hunk 3 above)
Comment (darkdestry-t) [reply src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java:60]: It isnt intended but has been the case for the original Name predicate. I think as a predicate its equal technically but shouldnt cause any issue if it isnt.

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 4]
```diff
@@ -9,26 +13,93 @@
...
-        assertParseSuccess(parser, "Alice Bob", expectedFindCommand);
+        FieldsContainsKeywordsPredicate predicateComparator =
+                new FieldsContainsKeywordsPredicate(Arrays.asList("Alice", "Reception"));
+
+        FindCommand expectedFieldsFindCommand = new FindCommand(predicateComparator, predicateComparator);
+        assertParseSuccess(parser, "Alice Reception", expectedFieldsFindCommand);
 
         // multiple whitespaces between keywords
-        assertParseSuccess(parser, " \n Alice \n \t Bob  \t", expectedFindCommand);
+        assertParseSuccess(parser, " \n Alice \n \t Reception  \t", expectedFieldsFindCommand);
```
[the remark this answers is not in the document]
Comment (darkdestry-t) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:102]: Non-issue. As comment says, white space is part of the test case.

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 5]
```diff
@@ -9,26 +13,93 @@
...
+    public void parse_validEmailArgs_returnsFindCommand() {
+        // no leading and trailing whitespaces
+        EmailContainsKeywordsPredicate predicateComparator =
+                new EmailContainsKeywordsPredicate(Arrays.asList("test@mail.com", "bob@box.net"));
+
+        FindCommand expectedEmailFindCommand = new FindCommand(predicateComparator, predicateComparator);
+        assertParseSuccess(parser, PREFIX_EMAIL + " test@mail.com bob@box.net", expectedEmailFindCommand);
+
+        // multiple whitespaces between keywords
+        assertParseSuccess(parser, PREFIX_EMAIL + " \n test@mail.com \n \t bob@box.net  \t", expectedEmailFindCommand);
```
[the remark this answers is not in the document]
Comment (darkdestry-t) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:89]: Fixed to put whitespace between prefix but not between keywords. The whitespace is part of test case

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 6]
```diff
@@ -9,26 +13,93 @@
...
+    public void parse_validRemarkArgs_returnsFindCommand() {
+        // no leading and trailing whitespaces
+        RemarkContainsKeywordsPredicate predicateComparator =
+                new RemarkContainsKeywordsPredicate(Arrays.asList("HR", "Network"));
+
+        FindCommand expectedRemarkFindCommand = new FindCommand(predicateComparator, predicateComparator);
+        assertParseSuccess(parser, PREFIX_REMARK + " HR Network", expectedRemarkFindCommand);
+
+        // multiple whitespaces between keywords
+        assertParseSuccess(parser, PREFIX_REMARK + " \n HR \n \t Network  \t", expectedRemarkFindCommand);
```
[the remark this answers is not in the document]
Comment (darkdestry-t) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:76]: Fixed to put whitespace between prefix but not between keywords. The whitespace is part of test case

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 7]
```diff
@@ -9,26 +13,93 @@
...
+    public void parse_validTagArgs_returnsFindCommand() {
+        // no leading and trailing whitespaces
+        TagContainsKeywordsPredicate predicateComparator =
+                new TagContainsKeywordsPredicate(Arrays.asList("Hotline", "Service"));
+
+        FindCommand expectedTagFindCommand = new FindCommand(predicateComparator, predicateComparator);
+        assertParseSuccess(parser, PREFIX_TAG + " Hotline Service", expectedTagFindCommand);
+
+        // multiple whitespaces between keywords
+        assertParseSuccess(parser, PREFIX_TAG + " \n Hotline \n \t Service  \t", expectedTagFindCommand);
```
[the remark this answers is not in the document]
Comment (darkdestry-t) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:63]: Fixed to put whitespace between prefix but not between keywords. The whitespace is part of test case

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 8]
```diff
@@ -9,26 +13,93 @@
...
+    public void parse_validNameArgs_returnsFindCommand() {
+        // no leading and trailing whitespaces
+        NameContainsKeywordsPredicate predicateComparator =
+                new NameContainsKeywordsPredicate(Arrays.asList("Alice", "Bob"));
+
+        FindCommand expectedNameFindCommand = new FindCommand(predicateComparator, predicateComparator);
+        assertParseSuccess(parser, PREFIX_NAME + " Alice Bob", expectedNameFindCommand);
+
+        // multiple whitespaces between keywords
+        assertParseSuccess(parser, PREFIX_NAME + " \n Alice \n \t Bob  \t", expectedNameFindCommand);
```
[the remark this answers is not in the document]
Comment (darkdestry-t) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:50]: Fixed to put whitespace between prefix but not between keywords. The whitespace is part of test case

**Review by yaowei-soc — APPROVED** [review]
LGTM

**Pull request merged**